server/proxy/pf_client.c

```diff
@@ -336,22 +507,19 @@ static BOOL pf_client_post_connect(freerdp* instance)
 	pf_client_register_update_callbacks(update);
 
 	/* virtual channels receive data hook */
-	client_receive_channel_data_original = instance->ReceiveChannelData;
+	pc->client_receive_channel_data_original = instance->ReceiveChannelData;
 	instance->ReceiveChannelData = pf_client_receive_channel_data_hook;
 
-	/* populate channel name -> channel ids map */
-	{
-		size_t i;
-		for (i = 0; i < config->PassthroughCount; i++)
-		{
-			char* channel_name = config->Passthrough[i];
-			UINT64 channel_id = (UINT64)freerdp_channels_get_id_by_name(instance, channel_name);
-			HashTable_Insert(pc->vc_ids, (void*)channel_name, (void*)channel_id);
-		}
-	}
-
 	instance->heartbeat->ServerHeartbeat = pf_client_on_server_heartbeat;
 
+	pc->connected = TRUE;
+
+	/* Send cached channel data */
+	ArrayList_Lock(pc->cached_server_channel_data);
```

Comment:
I was wondering, if this `ArrayList_Lock/Unlock` is required as `ArrayList_ForEach` does the locking internally  (although there's the `ArrayList_Clear` after ) ?

Reply:
yes, as otherwise a new entry could be added before the clear resulting in duplicate data sent.